cli/admin/call.go

```diff
@@ -0,0 +1,187 @@
+		log.Warnf("Admin function %q should return string or string array value", funcName)
+	}
+
+	return nil
+}
+
+func getFuncCallArgs(conn net.Conn, funcName string, flagSet *pflag.FlagSet, args []string) ([]FuncCallArg, error) {
+	helpResRawMap, err := getFuncHelpRawMap(funcName, conn)
+	if err != nil {
+		return nil, fmt.Errorf("Failed to get function %q signature: %s", funcName, err)
+	}
+
+	argsSpec, err := getArgsSpec(helpResRawMap)
+	if err != nil {
+		return nil, fmt.Errorf("Failed to get %q arguments spec: %s", funcName, err)
+	}
+
+	conflictingFlagNames := getConflictingFlagNames(argsSpec, flagSet)
+	if len(conflictingFlagNames) > 0 {
+		return nil, fmt.Errorf(
+			"Function has arguments with names that conflicts with `cartridge admin` flags: %s",
+			strings.Join(conflictingFlagNames, ", "),
+		)
+	}
+
+	funcCallArgs := make([]FuncCallArg, len(argsSpec))
+
+	i := 0
+	for argName, argSpec := range argsSpec {
+		funcCallArgs[i] = FuncCallArg{
+			Name: argName,
+			Type: argSpec.Type,
+		}
+
+		switch argSpec.Type {
+		case argTypeString:
+			flagSet.StringVar(&funcCallArgs[i].StringValue, argName, "", argSpec.Usage)
+		case argTypeNumber:
+			flagSet.IntVar(&funcCallArgs[i].NumberValue, argName, 0, argSpec.Usage)
+		case argTypeBool:
+			flagSet.BoolVar(&funcCallArgs[i].BoolValue, argName, false, argSpec.Usage)
+		default:
+			return nil, fmt.Errorf(
+				"Admin function %q accepts value of unsupported type: %s",
+				funcName, argSpec.Type,
+			)
+		}
+
+		i++
+	}
+
+	flagSet.SetNormalizeFunc(func(f *pflag.FlagSet, name string) pflag.NormalizedName {
+		return pflag.NormalizedName(normalizeFlagName(name))
+	})
+
+	flagSet.ParseErrorsWhitelist = pflag.ParseErrorsWhitelist{
+		UnknownFlags: false,
+	}
+	if err := flagSet.Parse(args); err != nil {
+		return nil, fmt.Errorf("Failed to parse %q function arguments: %s", funcName, err)
+	}
+
+	// set Changed for args
+	for i := range funcCallArgs {
+		flag := flagSet.Lookup(funcCallArgs[i].Name)
+		if flag == nil {
+			return nil, project.InternalError("Flag %q isn't found", funcCallArgs[i].Name)
+		}
+
+		funcCallArgs[i].Changed = flag.Changed
+	}
+
+	return funcCallArgs, nil
+}
+
+func serializeArgs(funcCallArgs []FuncCallArg) (string, error) {
+	argStrings := []string{}
+
+	for _, funcCallArg := range funcCallArgs {
+		var valueStr string
+
+		if !funcCallArg.Changed {
+			continue
+		}
+
+		switch funcCallArg.Type {
+		case argTypeString:
+			valueStr = fmt.Sprintf("'%s'", funcCallArg.StringValue)
+		case argTypeNumber:
+			valueStr = fmt.Sprintf("%d", funcCallArg.NumberValue)
+		case argTypeBool:
+			valueStr = fmt.Sprintf("%t", funcCallArg.BoolValue)
+		default:
+			return "", project.InternalError("Received argument with unsupported type: %s", funcCallArg.Type)
+		}
+
+		argStrings = append(argStrings, fmt.Sprintf("%s = %s", funcCallArg.Name, valueStr))
+	}
+
+	return fmt.Sprintf(`{ %s }`, strings.Join(argStrings, ", ")), nil
+}
+
+var (
+	adminCallFuncBodyTmpl = `
```

Comment:
Why is it not const?

Reply:
Because it's passed to `templates.GetTemplatedStr` function that accepts a pointer to string. I'm not sure that it's a good approach, but it should be fixed separately anyway.